### PR TITLE
Tpu save

### LINKED ISSUE #2303: Loading a model checkpoint that is trained on TPU using a GPU
#### What is your question?
Is it possible to load a model that is trained on a TPU saved using ModelCheckpoint on a GPU for inference?

#### Code

```
        model = LightModel(hparams)
        trainer = pl.Trainer(resume_from_checkpoint=str(ckpt), gpus=1)
        trainer.test(model)
```

#### What have you tried?## ❓ Questions and Help

Tried to normally load the weights as with a GPU but throws an error.

#### What's your environment?
Kaggle GPU
torchvision==0.6.0a0+82fd1c8
torch==1.5.0
pytorch-lightning-0.8.1

### LINKED ISSUE #2700: Checkpointing is broken on TPUs
## 🐛 Bug

Pytorch/XLA saves checkpoints using the following syntax which is not supported in pytorch-lightning.
```
import torch_xla.core.xla_model as xm
xm.save()
```
It is a little tricky to support because `xm.save()` has a barrier inside it and it checks for rank=0 while `torch.save` doesn't. This means `torch.save` should be called only on the process with rank=0 (which pytorch-lighting does) but `xm.save()` should be called by all processes (or it will wait forever at the barrier). This means pytorch-lightning code that checks for the rank ([here](https://github.com/PyTorchLightning/pytorch-lightning/blob/0.8.5/pytorch_lightning/callbacks/model_checkpoint.py#L265-L269) will need to be switched off on TPUs.

### To Reproduce

1. Train any model on TPUs using PyTorch/XLA with `ptl.Trainer(checkpoint_callback=[ModelCheckpoint(...)], num_tpu_cores=8)`
2. Wait until the model saves one checkpoint then kill the process
3. Try to load the saved checkpoint with `ptl.Trainer(resume_from_checkpoint='path_to_saved_checkpoint', num_tpu_cores=8)`

### Expected behavior

Loading checkpoint successfully. 

### Environment

pytorch-lightning==v0.8.5

### Additional Context

Thanks to @matt-peters for finding the bug and suggesting the solution mention below.

### LINKED ISSUE #3660: Colab TPU training stuck at end of epoch if checkpoint_callback=True
I am seeing training stuck at the last step of the 1st epoch when `checkpoint_callback=True` is enabled. I am passing `tpu_cores=8` to the trainer.

If it's `False` training is still slow for the first few steps and then it speeds up (I guess that this is expected since it takes a few steps for xla to compile stuff).

Here is a [link](https://colab.research.google.com/drive/1QCI_Nuz0iswNkVcdkJOFBtR8War1Xr7h?usp=sharing) to the colab notebook example.

```python

### Setup Model
class GPT2Tuned(pl.LightningModule):
    def __init__(self, hparams: argparse.Namespace):
        super().__init__()
        self.hparams = hparams
        self.model = AutoModelForCausalLM.from_pretrained(self.hparams.model_name_or_path)

    def forward(self, **inputs):
        return self.model(**inputs)
        
    def _step(self,batch):
        inputs = {"input_ids": batch[0]}
        inputs["labels"] = batch[0]
        outputs = self(**inputs)
        loss = outputs[0]
        return dict(loss=loss)

    def training_step(self, batch, batch_idx):
        out_dict = self._step(batch)
        return dict(loss=out_dict["loss"], log=out_dict)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

    def train_dataloader(self):
      # fake data
        dataset = TensorDataset(torch.randint(10000,(10000,128)).long()) 
        return  DataLoader(
            dataset,
            batch_size=self.hparams.batch_size,
            shuffle=True
        )

    @staticmethod
    def add_model_specific_args(parser):
        parser.add_argument("--model_name_or_path",type=str)
        parser.add_argument("--batch_size", default=4, type=int)
        parser.add_argument("--learning_rate", default=5e-5, type=float)
        return parser

#### Training 
parser = argparse.ArgumentParser()
parser = GPT2Tuned.add_model_specific_args(parser)
parser = pl.Trainer.add_argparse_args(parser)
args = parser.parse_args(["--gpus","0"])

args.__dict__.update(dict(model_name_or_path  = "gpt2-medium",
                          batch_size = 3,
                          precision=16,
                          tpu_cores=8))

model = GPT2Tuned(args)
trainer = pl.Trainer.from_argparse_args(args, checkpoint_callback=True)  
trainer.fit(model)

 ```

### PR DESCRIPTION
**Pull request by lezwon**
## What does this PR do?

Fixes #2700 
fixes #2303
fixes #3660

- Move the XLA tensors within a checkpoint to CPU before saving.
- Lazy check if TPU device exists
- Removed calling `accelerator.barrier()` on TPU when calling `.test()` as the multiprocessing begins only in `.fit()`
- Moved TPU device check to run when Trainer initialized.

Related to #3044 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/cloud_io.py [hunk 1]
```diff
@@ -42,6 +41,18 @@ def get_filesystem(path: pathlike):
...
+    if isinstance(item, dict):
+        new_dict = item.copy()
+        for key, val in item.items():
+            new_dict[key] = move_to_cpu(val)
+        return new_dict
+    if isinstance(item, list):
+        return list(map(move_to_cpu, item))
+    if isinstance(item, torch.Tensor):
+        return item.cpu()
+
```
Comment (ananthsub) [inline pytorch_lightning/utilities/cloud_io.py:54]: check out `move_data_to_device` here: https://github.com/PyTorchLightning/pytorch-lightning/blob/c7511bcbb49565c5b24f368fa25862f740f3ec28/pytorch_lightning/utilities/apply_func.py#L92

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/utilities/cloud_io.py (hunk 1 above)
Comment (lezwon) [reply pytorch_lightning/utilities/cloud_io.py]: aye thanks! Didn't know about this one 👍

**Review by williamFalcon — CHANGES_REQUESTED** [review]
@lezwon does that move to cpu really belong in this file? 
Why is this needed for TPUs?
You realize this will affect more than just TPUs no? this is why i don't like making these changes outside the accelerator.

If it turns out this is the correct place for this, please add a on_save() function to each accelerator. Then make all of them no op:

```python
def on_save(self, ...):
    pass
```

And add the change ONLY to the TPU accelerator.
```python
def on_save(self, ...):
   your_changes()
```

Going forward, any changes that are accelerator specific should not be done inside methods like this. Instead, each accelerator needs to implement this method and then called like:
```
self.accelerator.on_save()
```

The reason is that we are trying to break up all the underlying accelerator code so they are easier to debug and changes to an accelerator won't break all the others.

**Comment by lezwon** [comment]
> @lezwon does that move to cpu really belong in this file?
> Why is this needed for TPUs?
> You realize this will affect more than just TPUs no? this is why i don't like making these changes outside the accelerator.
> 
> If it turns out this is the correct place for this, please add a on_save() function to each accelerator. Then make all of them no op:
> 
> ```python
> def on_save(self, ...):
>     pass
> ```
> 
> And add the change ONLY to the TPU accelerator.
> 
> ```python
> def on_save(self, ...):
>    your_changes()
> ```
> 
> Going forward, any changes that are accelerator specific should not be done inside methods like this. Instead, each accelerator needs to implement this method and then called like:
> 
> ```
> self.accelerator.on_save()
> ```
> 
> The reason is that we are trying to break up all the underlying accelerator code so they are easier to debug and changes to an accelerator won't break all the others.

The XLA guide recommends users to move the tensors to CPU before saving so that they can be loaded on non-TPU devices. Lightning users have faced this issue wherein they have trained a model on a TPU and arent able to use it on a GPU or CPU.  Hence we need to move these tensors to CPU before saving. I get your point about separating such code and making it accelerator specific. Will refactor this into `on_save`. 👍

**Comment by SeanNaren** [comment]
hey @lezwon any activity here? were you able to reproduce this using the boring model btw? Don't mind picking this up

**Comment by lezwon** [comment]
hey @SeanNaren, I've been making some refactors to make this change TPU specific. WIll push an updated branch soon :)

**Comment by lezwon** [comment]
@Borda I've added a new tpu test file `tests/backends/test_tpu_backend.py`. Can we add that too to the CI tests? It's not running currently.

**Comment by Borda** [comment]
> @Borda I've added a new tpu test file `tests/backends/test_tpu_backend.py`. Can we add that too to the CI tests? It's not running currently.

sure, pls add the file here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/f63fec9323c319720d78b452e9fe84b97ce7644e/dockers/tpu-tests/tpu_test_cases.jsonnet#L24

**Comment by lezwon** [comment]
@Borda @williamFalcon Have done multiple changes and optimizations in this PR to get tests working. Do let me know if I should make a separate PR for these.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/xla_device_utils.py [hunk 2]
```diff
@@ -41,11 +40,12 @@ def wrapper(*args, **kwargs):
...
         proc = Process(target=inner_f, args=(queue, func, *args), kwargs=kwargs)
         proc.start()
-        proc.join(10)
+        proc.join(20)
         try:
             return queue.get_nowait()
         except q.Empty:
-            return False
+            traceback.print_exc()
+            return None
```
Comment (Borda) [inline pytorch_lightning/utilities/xla_device_utils.py:48]: not sure, would exist work be better?

File: tests/backends/test_tpu_backend.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+from pytorch_lightning import Trainer
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils as xla_utils
+from tests.base.boring_model import BoringModel
+from tests.base.develop_utils import pl_multi_process_test
+
+
+@pytest.mark.skipif(not xla_utils.tpu_device_exists(), reason="test requires TPU machine")
+@pl_multi_process_test
+def test_resume_training_on_cpu():
```
Comment (Borda) [inline tests/backends/test_tpu_backend.py:26]: all tests included training shall go to `tmpdir`

**Review by Borda — COMMENTED** [review]
some minor comments but it looks good to me...

**Comment by SeanNaren** [comment]
hey @lezwon let us know when this is ready to review :)

**Comment by lezwon** [comment]
@SeanNaren sure.. Just adding some minor changes with @Borda requested.

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -31,14 +31,12 @@
...
 
-TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists()
 try:
     from apex import amp
 except ImportError:
     amp = None
 
-if TPU_AVAILABLE:
-    import torch_xla
+if importlib.util.find_spec("torch_xla"):
```
[the remark this answers is not in the document]
Comment (lezwon) [reply pytorch_lightning/trainer/data_loading.py]: Cool.  Adding a method for that in XLAUtils.

File: pytorch_lightning/trainer/data_loading.py [hunk 5]
```diff
@@ -142,7 +140,7 @@ def replace_sampler(self, dataloader, sampler):
         return dataloader
 
     def _get_distributed_sampler(self, dataloader, shuffle):
-        if self.use_tpu:
+        if self.use_tpu and XLADeviceUtils.tpu_device_exists():
```
[the remark this answers is not in the document]
Comment (lezwon) [reply pytorch_lightning/trainer/data_loading.py]: Make a change that checks for tpu devices which setting `on_tpu`. That should let me remove this check from this position. :)

File: pytorch_lightning/utilities/xla_device_utils.py (hunk 2 above)
Comment (lezwon) [reply pytorch_lightning/utilities/xla_device_utils.py:48]: as in? I changed this to return `None` just to know when the tpu check timed out.

File: tests/backends/test_tpu_backend.py (hunk 3 above)
Comment (lezwon) [reply tests/backends/test_tpu_backend.py]: Made the change 👍

File: pytorch_lightning/utilities/xla_device_utils.py [hunk 6]
```diff
@@ -40,11 +40,12 @@ def wrapper(*args, **kwargs):
...
```
Comment (justusschock) [inline pytorch_lightning/utilities/xla_device_utils.py:48]:
```suggestion
            return
```

**Comment by lezwon** [comment]
@Borda I've made the changes you requested. Guess it's ready for review now.
cc: @SeanNaren

File: pytorch_lightning/trainer/data_loading.py [hunk 7]
```diff
@@ -11,25 +11,31 @@
...
 from pytorch_lightning.utilities.data import has_iterable_dataset, has_len
 from pytorch_lightning.utilities.debugging import InternalDebugger
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
 from pytorch_lightning.utilities.model_utils import is_overridden
+from pytorch_lightning.utilities.xla_device_utils import XLADeviceUtils
+
+try:
+    from apex import amp
+except ImportError:
+    amp = None
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/data_loading.py:36]: Why is this needed? @lezwon I think we can remove this right?

**Review by SeanNaren — COMMENTED** [review]
Just a small cleanup I think, but nice work. Good to ensure XLA/Device exists are separate! probably handles a lot of edge cases.

File: pytorch_lightning/trainer/data_loading.py (hunk 7 above)
Comment (lezwon) [reply pytorch_lightning/trainer/data_loading.py:36]: oops.. i think this appeared accidentally while merging and squashing commits. Will have some occurrences removed. :)

File: pytorch_lightning/accelerators/accelerator_connector.py [hunk 8]
```diff
@@ -350,7 +350,7 @@ def set_distributed_mode(self):
 
         rank_zero_info(f'GPU available: {torch.cuda.is_available()}, used: {self.trainer.on_gpu}')
         num_cores = self.trainer.tpu_cores if self.trainer.tpu_cores is not None else 0
-        rank_zero_info(f'TPU available: {XLA_AVAILABLE}, using: {num_cores} TPU cores')
+        rank_zero_info(f'TPU available: {XLADeviceUtils.tpu_device_exists()}, using: {num_cores} TPU cores')
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator_connector.py:353]: Why removing `XLA_AVAILABLE` ?

Comment (lezwon) [reply pytorch_lightning/accelerators/accelerator_connector.py:353]: `XLA_AVAILABLE` only checks if the XLA library is present. We need to ensure the TPU device exists on the machine :)

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1239,7 +1237,7 @@ def optimizer_step(self, epoch, batch_idx, optimizer, optimizer_idx,
             model hook don't forget to add the call to it before ``optimizer.zero_grad()`` yourself.
 
         """
-        if on_tpu:
+        if on_tpu and XLADeviceUtils.tpu_device_exists():
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1240]: doesn't on_tpu imply we already know device exists?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 10]
```diff
@@ -227,6 +227,9 @@ def hpc_save(self, folderpath: str, logger):
 
         model.on_hpc_save(checkpoint)
 
+        if self.trainer.accelerator_backend:
+            checkpoint = self.trainer.accelerator_backend.on_save(checkpoint)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:231]: isn't accelerator backend always set? even on cpu?

**Review by awaelchli — APPROVED** [review]
:)

File: pytorch_lightning/accelerators/accelerator_connector.py (hunk 8 above)
Comment (Borda) [reply pytorch_lightning/accelerators/accelerator_connector.py:353]: we have defined not `TPU_AVALAIBLE` but to the point, that you have installed XLA does not mean you also TPU device

File: tests/utilities/test_xla_device_utils.py [hunk 11]
```diff
@@ -16,40 +16,32 @@
...
 
 if XLA_AVAILABLE:
     import torch_xla.core.xla_model as xm
 
 
-@pytest.mark.skipif(XLA_AVAILABLE, reason="test requires torch_xla to be absent")
+@pytest.mark.skipif(TPU_AVAILABLE, reason="test requires torch_xla to be absent")
 def test_tpu_device_absence():
     """Check tpu_device_exists returns None when torch_xla is not available"""
     assert xla_utils.XLADeviceUtils.tpu_device_exists() is None
```
Comment (Borda) [inline tests/utilities/test_xla_device_utils.py:29]: this is tricky testing itself and for the testing set "if I am true test that I am true :D"

File: tests/utilities/test_xla_device_utils.py [hunk 12]
```diff
@@ -16,40 +16,32 @@
...
-from pytorch_lightning.utilities import XLA_AVAILABLE
+from pytorch_lightning.utilities import XLA_AVAILABLE, TPU_AVAILABLE
 from tests.base.develop_utils import pl_multi_process_test
 
 if XLA_AVAILABLE:
     import torch_xla.core.xla_model as xm
 
 
-@pytest.mark.skipif(XLA_AVAILABLE, reason="test requires torch_xla to be absent")
+@pytest.mark.skipif(TPU_AVAILABLE, reason="test requires torch_xla to be absent")
```
Comment (Borda) [inline tests/utilities/test_xla_device_utils.py:26]:
```suggestion
# lets hope that in or env we have installed XLA only for TPU devices, otherwise, it is testing in the cycle "if I am true test that I am true :D"
@pytest.mark.skipif(XLA_AVAILABLE, reason="test requires torch_xla to be absent")
```

**Review by Borda — APPROVED** [review]
lgtm

File: tests/utilities/test_xla_device_utils.py (hunk 11 above)
Comment (lezwon) [reply tests/utilities/test_xla_device_utils.py:30]: ah.. there seems to be a mismatch here from my local. I basically made this into
```python
@pytest.mark.skipif(not xla_utils.XLADeviceUtils.xla_available(), reason="test requires torch_xla to be installed")
```

Not sure how I pushed this code.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
Comment (lezwon) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:231]: I think it isnt set when `auto_lr_find` is used.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/core/optimizer.py [hunk 13]
```diff
@@ -11,38 +11,14 @@
...
-from torch.nn import Module
-from torch.optim import SGD
 from torch.optim.optimizer import Optimizer
 
-from pytorch_lightning import _logger as log
-from pytorch_lightning.callbacks import Callback
-from pytorch_lightning.core.grads import GradInformation
-from pytorch_lightning.core.hooks import CheckpointHooks, DataHooks, ModelHooks
-from pytorch_lightning.core.memory import ModelSummary
-from pytorch_lightning.core.saving import ALLOWED_CONFIG_TYPES, PRIMITIVE_TYPES, ModelIO
```
Comment (williamFalcon) [inline pytorch_lightning/core/optimizer.py:37]: where did this go? cc @tchaton

**Review by williamFalcon — APPROVED** [review]

File: pytorch_lightning/core/optimizer.py (hunk 13 above)
Comment (tchaton) [reply pytorch_lightning/core/optimizer.py:37]: I guess useless imports.

**Review by tchaton — APPROVED** [review]

File: pytorch_lightning/core/optimizer.py (hunk 13 above)
Comment (Borda) [reply pytorch_lightning/core/optimizer.py:37]: shall not be there at all in the first place, @SeanNaren
